huntsman/observatory.py

```diff
@@ -512,3 +363,159 @@ def _create_flat_field_observation(self, alt=None, az=None,
+            is_saturated = False
+            for cam_name, info in camera_events.items():
+                filename = info['filename']
+                self.logger.debug("Checking counts for {}".format(filename))
+
+                # Unpack fits if compressed
+                if not os.path.exists(filename) and \
+                        os.path.exists(filename.replace('.fits', '.fits.fz')):
+                    fits_utils.fpack(filename.replace('.fits', '.fits.fz'), unpack=True)
+
+                # Calculate average counts per pixel
+                data = fits.getdata(filename)
+                mean_counts, _, _ = stats.sigma_clipped_stats(data)
+                mean_counts -= bias
+                self.logger.debug("Counts: {}".format(mean_counts))
+                if (mean_counts < min_counts) or (mean_counts > max_counts):
+                    self.logger.debug("Counts outside min/max range, should be discarded.")
+
+                # Check if camera is saturated
+                if mean_counts >= max_counts:
+                    is_saturated = True
+
+                # Get suggested exposure time
+                elapsed_time = (utils.current_time() - start_time).sec
+                self.logger.debug("Elapsed time: {}".format(elapsed_time))
+                exptime = int(exptimes[cam_name][-1].value * (target_adu / mean_counts) *
+                              (2.0 ** (elapsed_time / 180.0)) + 0.5)
+                exptime = max(1, exptime)
+                self.logger.debug("Suggested exptime for {}: {}".format(cam_name, exptime))
+                exptimes[cam_name].append(exptime * u.second)
+
+            # Stop flats if any time is greater than max
+            if any([t[-1].value >= max_exptime for t in exptimes.values()]) & \
+                    (not self.past_midnight):
+                self.logger.debug("Exposure times greater than max, stopping flat fields.")
+                break
+
+            # Stop flats if we have taken too many.
+            if any([len(t) >= max_num_exposures for t in exptimes.values()]):
+                self.logger.debug("Too many flats, quitting.")
+                break
+
+            # Handle saturated cameras
+            if is_saturated and exptimes[cam_name][-1].value < 2:
+                if self.past_midnight:
+                    self.logger.debug('Saturated with short exposure. Stopping flat fields.')
+                    break
+                else:
+                    self.logger.debug('Saturated with short exposure. Waiting 30 seconds'
+                                      'before next exposure.')
+                    max_num_exposures += 1
+                    time.sleep(30)
+
+        # Add a bias exposure
+        for cam_name in camera_names:
+            exptimes[cam_name].append(0 * u.second)
+
+        # Take darks for each exposure we took
+        for i in range(exposure_count):
+            _exptimes = {cam_name: exptimes[cam_name][i] for cam_name in camera_names}
```

Comment:
This should be checking if the camera has the ability to take darks. Or we should skip this whole block if we know at the observation level.

Reply:
@wtgee what is the correct way to check if a camera can take darks?

Reply:
There isn't really any way to do this at the moment. Might need to add a `Camera` property for that.